py_ecc/bls/api.py

```diff
@@ -109,3 +113,48 @@ def verify_multiple(pubkeys: Sequence[BLSPubkey],
         return final_exponentiation == FQ12.one()
     except (ValidationError, ValueError, AssertionError):
         return False
+
+
+def verify_multiple_multiple(
+        signatures: Sequence[BLSSignature],
+        pubkeys_and_messages: Tuple[Sequence[BLSPubkey], Sequence[Hash32]],
+        domain: int)-> bool:
+    if len(signatures) != len(pubkeys_and_messages):
+        raise ValidationError(
+            "len(signatures) (%s) should be equal to len(pubkeys_and_messages) (%s)" % (
+                len(signatures), len(pubkeys_and_messages)
+            )
+        )
+    try:
+        random_ints = (1,) + tuple(2**randbelow(64) for _ in range(len(signatures) - 1))
+        o = FQ12.one()
+        for index, pm in enumerate(pubkeys_and_messages):
+            pubkeys, message_hashes = pm  # type: ignore
+            len_msgs = len(message_hashes)
+            if len(pubkeys) != len_msgs:
+                raise ValidationError(
+                    "len(pubkeys) (%s) should be equal to len(message_hashes) (%s)" % (
+                        len(pubkeys), len_msgs
+                    )
+                )
+            for m_pubs in set(message_hashes):
+
+                # aggregate the pubs
+                group_pub = Z1
+                for i in range(len_msgs):
+                    if message_hashes[i] == m_pubs:
```

Comment:
IIUC this code is finding signatures with common public keys and grouping them.  The current implementation seems to have `~O(n ** 2)` complexity since it loops over the hashes once for each hash.  You should be able to reduce this using something like:

- use `cytoolz.groupby(operator.itemgetter(1), enumerate(message_hashes))` to group the message hashes and retain their indices.
- loop over that data structure once, pulling out the indices to produce the `group_pub` for each set of common hashes.